File: internal/breakdown/breakdown.go
// Package breakdown uses an LLM to break down a goal into a structured task list.
package breakdown

import (
	"context"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/jasjeetmavi/orca/internal/driver"
	"github.com/jasjeetmavi/orca/internal/explore"
	"github.com/jasjeetmavi/orca/internal/interaction"
	"github.com/jasjeetmavi/orca/internal/worker"
	"github.com/jasjeetmavi/orca/prompts"
)

type ProposedTask struct {
	Title            string `json:"title"`
	Description      string `json:"description"`
	DependsOnIndices []int  `json:"depends_on_indices"`
	SuggestedTool    string `json:"suggested_tool"`
}

type Breaker struct {
	toolName     string
	driver       driver.Driver
	model        string
	timeout      time.Duration
	repoDir      string
	interactions *interaction.Store
}

func New(toolName string, d driver.Driver, model string, timeout time.Duration, repoDir string, interactions ...*interaction.Store) *Breaker {
	var store *interaction.Store
	if len(interactions) > 0 {
		store = interactions[0]
	}
	return &Breaker{toolName: toolName, driver: d, model: model, timeout: timeout, repoDir: repoDir, interactions: store}
}

func (b *Breaker) Run(taskID *string, goal string) ([]ProposedTask, string, error) {
	contextSection := ""
	if ctx := explore.LoadContext(b.repoDir); ctx != "" {
		contextSection = "## Codebase Context\n\n" + ctx + "\n\n"
	}
	prompt := fmt.Sprintf(prompts.Breakdown, contextSection, goal)

	adapter := worker.NewAdapter(b.driver, b.model, b.timeout)

	interactionID := ""
	result, err := interaction.RunWithTracking(
		b.interactions,
		taskID,
		interaction.PhaseBreakdown,
		b.toolName,
		adapter,
		func() (*worker.Result, error) {
			return adapter.Execute(context.Background(), interaction.PhaseBreakdown, prompt, b.repoDir)
		},
		interaction.WithOnBegin(func(w *interaction.Writer) {
			interactionID = w.ID()
		}),
		interaction.WithFinishFn(func(result *worker.Result, runErr error) (string, []interaction.FinishOption) {
			status := "completed"
			opts := []interaction.FinishOption{}
			if result != nil {
				opts = append(opts, interaction.WithCost(result.InputTokens, result.OutputTokens, result.TotalCost))
			}
			if runErr != nil {
				status = "failed"
				opts = append(opts, interaction.WithError(runErr.Error()))
			} else if result == nil || result.ExitCode != 0 {
				status = "failed"
				exitCode := -1
				stderr := ""
				if result != nil {
					exitCode = result.ExitCode
					stderr = result.Stderr
				}
				opts = append(opts, interaction.WithError(fmt.Sprintf("breaker exited %d: %s", exitCode, stderr)))
			}
			return status, opts
		}),
	)
	stdout := ""
	exitCode := -1
	stderr := ""
	if result != nil {
		stdout = result.Stdout
		exitCode = result.ExitCode
		stderr = result.Stderr
	}
	if err != nil {
		return nil, interactionID, fmt.Errorf("execute breaker: %w", err)
	}
	if exitCode != 0 {
		return nil, interactionID, fmt.Errorf("breaker exited %d: %s", exitCode, stderr)
	}

	output := stdout
	tasks, err := parseProposedTasks(output)
	if err != nil {
		return nil, interactionID, fmt.Errorf("parse task breakdown: %w\nraw output:\n%s", err, output)
	}
	return tasks, interactionID, nil
}

var (
	taskHeaderRe    = regexp.MustCompile(`(?i)^###\s*task\s+(\d+)(?::\s*(.*))?$`)
	listFieldLineRe = regexp.MustCompile(`^-+\s*([A-Za-z ]+)\s*:\s*(.*)$`)
)

func parseProposedTasks(output string) ([]ProposedTask, error) {
	return parseMarkdownProposedTasks(output)
}

func parseMarkdownProposedTasks(output string) ([]ProposedTask, error) {
	lines := strings.Split(strings.TrimSpace(output), "\n")
	if len(lines) == 0 {
		return nil, fmt.Errorf("empty response")
	}

	tasks := make([]ProposedTask, 0, 8)
	current := ProposedTask{}
	inTask := false
	appendCurrent := func() {
		if !inTask {
			return
		}
		current.Title = strings.TrimSpace(current.Title)
		current.Description = strings.TrimSpace(current.Description)
		if current.Title == "" && current.Description == "" {
			return
		}
		if current.Title == "" {
			current.Title = "Task"
		}
		if current.Description == "" {
			current.Description = current.Title
		}
		current.SuggestedTool = normalizeSuggestedTool(current.SuggestedTool)
		current.DependsOnIndices = normalizeDependsOn(current.DependsOnIndices)
		tasks = append(tasks, current)
	}

	for _, raw := range lines {
		line := strings.TrimSpace(raw)
		if line == "" {
			continue
		}

		if m := taskHeaderRe.FindStringSubmatch(line); m != nil {
			appendCurrent()
			current = ProposedTask{}
			inTask = true
			if len(m) > 2 && strings.TrimSpace(m[2]) != "" {
				current.Title = strings.TrimSpace(m[2])
			}
			continue
		}
		if !inTask {
			continue
		}

		if m := listFieldLineRe.FindStringSubmatch(line); m != nil {
			key := strings.ToLower(strings.TrimSpace(m[1]))
			value := strings.TrimSpace(m[2])
			switch key {
			case "title":
				current.Title = value
			case "description":
				current.Description = value
			case "depends on":
				current.DependsOnIndices = parseDependsOnNumbers(value)
			case "suggested tool":
				current.SuggestedTool = value
			}
		}
	}
	appendCurrent()

	if len(tasks) == 0 {
		return nil, fmt.Errorf("no tasks found in markdown response")
	}
	return normalizeProposedTasks(tasks), nil
}

func parseDependsOnNumbers(raw string) []int {
	clean := strings.TrimSpace(strings.ToLower(raw))
	if clean == "" || clean == "none" || clean == "n/a" {
		return nil
	}
	parts := strings.FieldsFunc(clean, func(r rune) bool {
		return r == ',' || r == ' ' || r == ';'
	})
	out := make([]int, 0, len(parts))
	for _, part := range parts {
		n, err := strconv.Atoi(strings.TrimSpace(part))
		if err != nil || n <= 0 {
			continue
		}
		out = append(out, n-1) // markdown template is 1-based
	}
	return out
}

func normalizeSuggestedTool(tool string) string {
	switch strings.ToLower(strings.TrimSpace(tool)) {
	case "claude", "codex":
		return strings.ToLower(strings.TrimSpace(tool))
	default:
		return ""
	}
}

func normalizeDependsOn(deps []int) []int {
	if len(deps) == 0 {
		return nil
	}
	seen := make(map[int]struct{}, len(deps))
	out := make([]int, 0, len(deps))
	for _, dep := range deps {
		if dep < 0 {
			continue
		}
		if _, ok := seen[dep]; ok {
			continue
		}
		seen[dep] = struct{}{}
		out = append(out, dep)
	}
	return out
}

func normalizeProposedTasks(tasks []ProposedTask) []ProposedTask {
	out := make([]ProposedTask, 0, len(tasks))
	for _, task := range tasks {
		title := strings.TrimSpace(task.Title)
		description := strings.TrimSpace(task.Description)
		if title == "" && description == "" {
			continue
		}
		if title == "" {
			title = "Task"
		}
		if description == "" {
			description = title
		}
		out = append(out, ProposedTask{
			Title:            title,
			Description:      description,
			DependsOnIndices: normalizeDependsOn(task.DependsOnIndices),
			SuggestedTool:    normalizeSuggestedTool(task.SuggestedTool),
		})
	}
	return out
}

File: server/src/domain/breakdown.ts
import type { Config } from '../db/schema';
import type { ToolPluginRegistry } from '../plugin/registry';
import { createInteractionRunner } from '../shared/interaction-runner';
import type { InteractionStore } from '../store/interactions';
import type { ProposedTask } from '../types/api';
import { runTool } from '../worker/worker';
import { extractJSONArray } from './llm';

const TASK_HEADER_RE = /^###\s*task\s+\d+(?::\s*(.*))?$/i;
const FIELD_RE = /^-+\s*([A-Za-z ]+)\s*:\s*(.*)$/;

interface RunBreakdownInput {
  repoDir: string;
  config: Config;
  registry: ToolPluginRegistry;
  interactions: InteractionStore;
  goal: string;
  taskID?: string;
  memoryContext?: string;
  toolOverride?: string;
  modelOverride?: string;
}

interface RunBreakdownResult {
  proposed: ProposedTask[];
  interactionId: string;
  tool: string;
  model: string;
}

export async function runBreakdown(
  input: RunBreakdownInput,
): Promise<RunBreakdownResult> {
  const runInteraction = await createInteractionRunner({
    config: input.config,
    registry: input.registry,
    repoDir: input.repoDir,
    interactions: input.interactions,
    runTool,
  });

  const contextSection = input.memoryContext?.trim()
    ? `## Retrieved Context\n\n${input.memoryContext.trim()}\n`
    : '';

  const {
    result: proposed,
    interactionId,
    tool,
    model,
  } = await runInteraction(
    {
      taskId: input.taskID?.trim() || null,
      type: 'breakdown',
      promptName: 'breakdown',
      promptArgs: [contextSection, input.goal.trim()],
      toolOverride: input.toolOverride ?? '',
      modelOverride: input.modelOverride ?? '',
      resolveErrorMessage: 'unable to resolve breakdown tool/model',
      exitErrorLabel: 'breakdown',
    },
    (output) => {
      const parsed = parseProposedTasks(output);
      if (parsed.length === 0) {
        throw new Error('no subtasks found in breakdown output');
      }
      return parsed;
    },
    (parsed) => ({
      output: JSON.stringify({
        result: 'proposed',
        data: { goal: input.goal, proposed: parsed },
      }),
    }),
  );

  return {
    proposed,
    interactionId,
    tool,
    model,
  };
}

export function normalizeProposedTasks(tasks: ProposedTask[]): ProposedTask[] {
  return tasks.map((task) => ({
    title: task.title?.trim() || 'Untitled task',
    description: task.description?.trim() || '',
    dependsOnIndices: Array.isArray(task.dependsOnIndices)
      ? task.dependsOnIndices.filter(
          (index) => Number.isInteger(index) && index >= 0,
        )
      : [],
    suggestedTool: task.suggestedTool?.trim() || 'claude',
  }));
}

function parseProposedTasks(output: string): ProposedTask[] {
  const parsedJSON = extractJSONArray<ProposedTask>(output);
  if (parsedJSON && parsedJSON.length > 0) {
    return normalizeProposedTasks(parsedJSON);
  }

  const lines = output.trim().split('\n');
  if (lines.length === 0) return [];

  const tasks: ProposedTask[] = [];
  let current: ProposedTask | null = null;

  const pushCurrent = () => {
    if (!current) return;
    if (!current.title.trim() && !current.description.trim()) {
      current = null;
      return;
    }
    tasks.push(current);
    current = null;
  };

  for (const rawLine of lines) {
    const line = rawLine.trim();
    if (!line) continue;

    const headerMatch = line.match(TASK_HEADER_RE);
    if (headerMatch) {
      pushCurrent();
      current = {
        title: String(headerMatch[1] ?? '').trim(),
        description: '',
        dependsOnIndices: [],
        suggestedTool: 'claude',
      };
      continue;
    }
    if (!current) continue;

    const fieldMatch = line.match(FIELD_RE);
    if (!fieldMatch) continue;

    const key = String(fieldMatch[1] ?? '')
      .trim()
      .toLowerCase();
    const value = String(fieldMatch[2] ?? '').trim();
    if (key === 'title') {
      current.title = value;
      continue;
    }
    if (key === 'description') {
      current.description = value;
      continue;
    }
    if (key === 'depends on') {
      current.dependsOnIndices = parseDependsOnNumbers(value);
      continue;
    }
    if (key === 'suggested tool') {
      current.suggestedTool =
        normalizeSuggestedTool(value) || current.suggestedTool;
    }
  }
  pushCurrent();

  return normalizeProposedTasks(tasks);
}

function parseDependsOnNumbers(raw: string): number[] {
  const cleaned = raw.trim().toLowerCase();
  if (!cleaned || cleaned === 'none' || cleaned === 'n/a') return [];

  return [
    ...new Set(
      cleaned
        .split(/[,\s;]+/g)
        .map((part) => Number.parseInt(part.trim(), 10))
        .filter((value) => Number.isFinite(value) && value > 0)
        .map((value) => value - 1),
    ),
  ];
}

function normalizeSuggestedTool(raw: string): string {
  const value = raw.trim().toLowerCase();
  if (value === 'claude' || value === 'codex') return value;
  if (value === 'none') return '';
  return '';
}
